### DISCUSSION TITLE
User/Module Lifecycle

### DISCUSSION THREAD

**Discussion by psychemedia** [mmh352/ou-container-builder#29] [General]
This came up in TM351 discussion last week:

- how long will a service remain available to students? Forum access and access to a VLE site expires after a certain period; so:
  - how long will students have access to a compute environment after a module presentation?
  - how long will students have access to their storage / files?
  - students should have access to both at least until any results / resubmit / retake /  appeal process has expired;
- if modules update images presentation on presentation, what image should a student who deferred use?
- some students use module software for project modules: how will that work?
  - How will students access files and maybe the compute environment they used a couple of years ago for their final project?

Note that students should always be able to run the computer environment via a local docker install, so it may be that we don't offer any long term guarantees and that local use is the only option after a certain period of time. Which might mean we need to maintain archive-retrievable copies of images too (I think DockerHub will increasingly look for more ways to expire not-recently-pulled free hosted images).

**Comment by mmh352** [comment]
* The first bullet point is essentially a set of management-level questions. They don't impact on the actual technical setup. My own view is that students should have access to the compute environment at least until they have a final mark for that module. 
* The second bullet point should follow the same rules as for module content and assessment. If deferred students re-start with the next presentation's content/assessment, then they should have access to the new image. If not, then they should continue to use the old image.
* This is essentially the same as the first bullet point. It is a management decisions for how long they should have access.
  * The sub-point on files is a separate issue. My own view is that the students should have access to their files until they complete their studies. Access can easily be provided via a simple, minimally resourced FileManager container.

**Comment by psychemedia** [comment]
Re: not impacting in the technical set up, yes and no? It could in part depend on how  access and auth works, and the extent to which whoever is looking after user registration and permissions on the hub has to actively manage adding or deleting users or their access to particular images.

One crude rule of thumb, if students access the hub from a link on the VLE, is to say that students should have access to the compute environment as long as they have access to the VLE materials (and the hub link therefrom); another might be that compute is like forums: at a certain point forums may become readonly, or closed (I'm not sure what the policy is) even if VLE learning content access continues, so we might limit access to compute to the period students have write access to forums etc.

If students are in affinity groups (eg by module and presentation) then access might be simplified accordingly (students on this module and that presentation get access to that module etc.)

The practicalities of user management / account creation generally is an issue in itself, eg if you have to enrol 1k students and 50 tutors per presentation, help with lost passwords etc etc. (Managing ALs and module teams raises its own concerns, not least in terms of how many environments they may need to access, and over what period; this may include things like overlapping presentations of the same module etc, but the `module-namespace` separation should address a lot of that sort of concern.)

Policy discussion, and technical constraints or practical management side effect, are also a discussion that makes to sense to have early, figuring out what positions a possible, reasonable, ridiculous etc, rather than getting to the end of a presentation and then suddenly having to make up policy on the fly, or having a sysop delete everything except the tape backup not realising that for some reason students should still have access even though the module end date was yesterday.

Jupyter stuff we've been using for years, but in many cases the recent flurry of adoption of various related things has been naive and not considered. A lot of things folk can discover via highly ranked web search results relate to dated advice, and does not represent current and emerging best practice (if there is any!), let alone roadmap solutions that are currently in development or on the further horizon (which may well be sooner than current production module first presentation dates, such are OU time scales.

An example of a conversation we aren't really having anywhere is the whole classic notebook thing. IT's reticence at looking at various solutions that already exist and are deployed at scale that allow for sensible provision of GPU powered kernels is another. Current and under development solutions for collaborative notebook working (either for team working or for AL support) are a third. Etc.

**Reply by mmh352** [reply]
I'm starting the process of seeing whether we can hook into the single-sign-on. Then all that needs to be done is to authorise students for specific containers. We'll see how that goes, but obviously there will have to be a UI for that.

I agree that policy discussions take time and policy impacts technology/process choices. However, unless unavoidable, policy should drive tech/process and not the other way around, which is why it is not my priority.

I agree that the deprecation of the classic notebook is an issue that is floating under the surface. The problem is that I don't use it, so I sort-of don't care too much :-D.
